File: python/src/resp_bench/command/factory.py
"""Factory mapping command names to implementations."""

from __future__ import annotations

from typing import Dict, List, Type

from ..config.command_config import CommandConfig
from .command import Command
from .impl.get_command import GetCommand
from .impl.ping_command import PingCommand
from .impl.set_command import SetCommand


class CommandFactory:
    _COMMAND_CLASSES: Dict[str, Type[Command]] = {
        "ping": PingCommand,
        "get": GetCommand,
        "set": SetCommand,
        # Future: hget, hset, lpush, lpop, sadd, smembers
    }

    @classmethod
    def create(cls, config: CommandConfig) -> Command:
        command_class = cls._COMMAND_CLASSES.get(config.command)
        if command_class is None:
            raise ValueError(
                f"Unknown command: {config.command}. "
                f"Supported: {', '.join(cls._COMMAND_CLASSES)}"
            )
        return command_class(config)

    @classmethod
    def create_all(cls, configs: List[CommandConfig]) -> List[Command]:
        return [cls.create(c) for c in configs]

    @classmethod
    def supported_commands(cls) -> List[str]:
        return list(cls._COMMAND_CLASSES.keys())

File: php/src/Command/Factory.php
<?php

declare(strict_types=1);

namespace RespBench\Command;

use InvalidArgumentException;
use RespBench\Command\Impl\GetCommand;
use RespBench\Command\Impl\PingCommand;
use RespBench\Command\Impl\SetCommand;
use RespBench\Config\CommandConfig;

/**
 * Factory for creating command instances.
 */
final class Factory
{
    /** @var array<string,class-string<Command>> */
    private const COMMAND_CLASSES = [
        'ping' => PingCommand::class,
        'get' => GetCommand::class,
        'set' => SetCommand::class,
    ];

    public static function create(CommandConfig $config): Command
    {
        $class = self::COMMAND_CLASSES[$config->command] ?? null;
        if ($class === null) {
            throw new InvalidArgumentException(
                "Unknown command: {$config->command}. Supported: "
                . implode(', ', array_keys(self::COMMAND_CLASSES))
            );
        }

        return new $class($config);
    }

    /**
     * @param list<CommandConfig> $configs
     * @return list<Command>
     */
    public static function createAll(array $configs): array
    {
        return array_map(static fn (CommandConfig $c): Command => self::create($c), $configs);
    }

    /**
     * @return list<string>
     */
    public static function supportedCommands(): array
    {
        return array_keys(self::COMMAND_CLASSES);
    }
}
